### PR TITLE
PI Backend V2.0

### PR DESCRIPTION
**Pull request by EtSubas**
fix DB subscriptions and writing to DB 
finish writing Serial handler
add actual multiprocessing to backend

### CODE REVIEW

File: backend/src/SerialHandler.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        self.send_message_workq = message_handler_workq
+        
+        # Open serial serial port
+        try:
+            self.serial_port = serial.Serial(port=port, baudrate=baudrate, bytesize=8, parity=serial.PARITY_NONE, timeout=None, stopbits=serial.STOPBITS_ONE)
+        except Exception:
+            self.serial_port = None
+            logger.error(f"Failed to open {thread_name} serial port {port}")
+            return
+        logger.success(f"Successfully opened {thread_name} serial port {port}")
```
Comment (cjchanx) [inline backend/src/SerialHandler.py:67]: We could also potentially have a Statuses table instead of an error table to put a little green icon thing on things like "Serial Radio / Serial RCU" or "DB Thread"

File: backend/src/SerialHandler.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    while (1):
+        # If there is any workq messages, process them first
+        # then once the queue is empty read the serial port
+        if not serial_workq.empty():
+            if not process_serial_workq_message(serial_workq.get(), ser_han):
+                return
+        else:
+            ser_han.handle_serial_message()
+
+        time.sleep(0.1)
```
Comment (cjchanx) [inline backend/src/SerialHandler.py:260]: This sleep 0.1 might be too much of a delay (100ms between each serial message read, correct me if wrong) -- while PySerial is buffered (which is probably why there aren't too many lost messages with just the RCU) it technically wouldn't keep up with a rate of > 10 messages per second right now.

There's also a weird dynamic here where read_until blocks until an incoming message is processed so our commands to write serial would probably stop working if we stopped receiving data for some reason.

One way to deal with this might be to use threading inside this process to make one separate WorkQ thread that just blocks forever for a workQ message, and the serial Rx thread which will block in read_until for a serial message, so both Tx/Rx will work at the rate that messages come in, and we won't have to poll at a fixed rate.

File: backend/src/main.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        # Initialize the database handler thread
+        db_thread = mp.Process(target=database_thread, args=('database', db_workq, message_handler_workq))
+        # Add the database thread to the thread pool
+        thread_pool['database'] = {'thread': db_thread, 'workq': db_workq}
+        
+        utl.thread_pool = thread_pool
+        return
+
+if __name__ == "__main__":
+  utl()
```
Comment (cjchanx) [inline backend/src/main.py:52]: Maybe add a quick argparse, or just a variable we can change here so we can change the logging level. You can implement this as a function in CommonLogger for example called `setup_logger(level)` which runs `logger.add()` with a different log level based on the one that's given.

We could also potentially add a function like `change_log_level()` that just removes and adds back the logger with a different level, but we probs wouldn't use this

File: backend/src/Utils.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    logger.error(f"Attempting to send message to non-existent thread: {serial_thread}")
+                    continue
+                Utils.thread_pool[serial_thread]['workq'].put(message)
+        else:
+            if (message.dest_thread not in Utils.thread_pool) or (not Utils.thread_pool[message.dest_thread]['thread'].is_alive()):
+                logger.error(f"Attempting to send message to non-existent thread: {message.dest_thread}")
+                return
+            dest_workq = Utils.thread_pool[message.dest_thread]['workq']
+            if dest_workq:
+                dest_workq.put(message)
```
Comment (cjchanx) [inline backend/src/Utils.py:99]: I feel like this part of Utils has become a ThreadManager kinda but it's alright to keep it like this for now

**Review by cjchanx — CHANGES_REQUESTED** [review]

File: backend/src/DatabaseHandler.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        collection to wait for commands created in the front end. 
+        The handler can also send telemetry data to the database
+        to be read by the front end.
+        """
+        logger.info("DatabaseHandler initializing")
+        DatabaseHandler.thread_workq = thread_workq
+        DatabaseHandler.send_message_workq = message_handler_workq
+        DatabaseHandler.thread_name = thread_name
+        DatabaseHandler.client = Client("http://10.13.141.121:8090/")
+        DatabaseHandler.client.collection('CommandMessage').subscribe(DatabaseHandler._handle_command_callback)
```
Comment (Jgerbrandt) [inline backend/src/DatabaseHandler.py:29]: Need to make sure this is the correct ip.

File: backend/src/DatabaseHandler.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    """
+
+    DatabaseHandler(thread_name, db_workq, message_handler_workq)
+
+    while 1:
+        # If there is any workq messages, process them
+        if not db_workq.empty():
+            if not process_workq_message(db_workq.get()):
+                return
+        time.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (EtSubas) [reply backend/src/DatabaseHandler.py:88]: makes sense love the verbose 😎
and yeah guess its just waiting for stuff to do so no sleep needed

Done 🫡

File: backend/src/SerialHandler.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        self.thread_name = thread_name
+        self.send_message_workq = message_handler_workq
+        
+        # Open serial serial port
+        try:
+            self.serial_port = serial.Serial(port=port, baudrate=baudrate, bytesize=8, parity=serial.PARITY_NONE, timeout=None, stopbits=serial.STOPBITS_ONE)
+        except Exception:
+            self.serial_port = None
+            logger.error(f"Failed to open {thread_name} serial port {port}")
+            return
```
[the remark this answers is not in the document]
Comment (EtSubas) [reply backend/src/SerialHandler.py:66]: sounds good ill make a note in the read me for this

File: backend/src/SerialHandler.py (hunk 1 above)
Comment (EtSubas) [reply backend/src/SerialHandler.py:67]: sounds good added a note about this too

File: backend/src/main.py (hunk 3 above)
Comment (EtSubas) [reply backend/src/main.py:52]: I like this idea andd think it should be added or something like it, however there is a problem with this since we are using mp. I realized that if something isn't staticly allocated the threads will use whatever the static value is.

For example I was trying to dynamically turn off an on logging to terminal from main (which would then go into common logger and either turn off or on logging to terminal) but then when we create the othere threads they do not look at what the main thread had already done to the logger andd just use the static version which was default connected to terminal and default dont add to the log fille.

A way that this could actually be fixed is by adding that backend handler thread which can work on making sure the logs are all handled in on place... maybe I'm worried about it getting overeloaded with too many logs but when we are out of the development portion of the backend we can remove a ton of the super super verbose logs maybe

TLDR: I think we should add this but not in this commit I'll make another todo in the readme and start the overhall after this merge.

File: backend/src/Utils.py (hunk 4 above)
Comment (EtSubas) [reply backend/src/Utils.py:99]: yeah I agree it has strayed from its original purpose.. 

I've split thee two into a utils and thread manager file, makes much more sense and going forward will be easier to expand.

File: backend/src/DatabaseHandler.py (hunk 5 above)
Comment (EtSubas) [reply backend/src/DatabaseHandler.py:29]: what is the correct IP? i added a thing to the readme to make sure this is generalized for front end and backend but If you have to IP I can put it in now

File: backend/src/main.py (hunk 3 above)
Comment (cjchanx) [reply backend/src/main.py:52]: sounds good, another idea that may work good for this is to just have a Configuration.py with global constants such as that DISABLE_LOGGING and log level cause that way we _should_ only ever have to touch one file to change how it's workin but yea i'm good with doing it after merge

File: backend/src/DatabaseHandler.py (hunk 5 above)
Comment (Jgerbrandt) [reply backend/src/DatabaseHandler.py:29]: We don't know yet. We'll have to setup the soar router and connect and see. This was just a todo/make sure we update it at some point.

File: backend/src/main.py (hunk 3 above)
Comment (EtSubas) [reply backend/src/main.py:52]: yeah that would be an easier fix plus can work for other constants like uart buads, ports, db url etc..

File: backend/src/DatabaseHandler.py (hunk 5 above)
Comment (EtSubas) [reply backend/src/DatabaseHandler.py:29]: sounsd good its ddone

File: backend/src/DatabaseHandler.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    The main loop of the database handler. It subscribes to the CommandMessage collection
+    """
+
+    DatabaseHandler(thread_name, db_workq, message_handler_workq)
+
+    while 1:
+        # If there is any workq messages, process them
+        if not db_workq.empty():
+            if not process_workq_message(db_workq.get(block=True)):
+                return
```
Comment (cjchanx) [inline backend/src/DatabaseHandler.py:87]: Make sure to also get rid of the `db_workq.empty()` check, because this is what'll make it basically spam-poll the queue as opposed to actually blocking on the .get call

Comment (EtSubas) [reply backend/src/DatabaseHandler.py:87]: Oh right I didnt know what you meant before but I get it, is this ok with the subscribe function? will the callback still be able to let us know there is a dataabase notification??

Comment (cjchanx) [reply backend/src/DatabaseHandler.py:87]: Subscribe should be ok, we should give it a quick test but based on a quick look at their code their event loops work with `anyio`, and we're not calling into any handlers so blocking here should be a-ok (maybe better for subscribe since that gives the subscribe function more time to run)

File: backend/src/main.py (hunk 3 above)
Comment (EtSubas) [reply backend/src/main.py]: I tested it with the blocking and seems to work fine

File: backend/src/DatabaseHandler.py (hunk 8 above)
Comment (EtSubas) [reply backend/src/DatabaseHandler.py]: idk what i just replied to.. but I tested this and it worked all good

File: backend/src/main.py (hunk 3 above)
Comment (EtSubas) [reply backend/src/main.py]: oh i diddnt do this but i addded a todo for the logs stuff

File: backend/src/main.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+
+# General imports =================================================================================
+import os, sys, git
+import multiprocessing as mp
+
+# Project specific imports ========================================================================
+git_repo = git.Repo(__file__, search_parent_directories=True).git.rev_parse("--show-toplevel")
+sys.path.insert(0, os.path.join(git_repo, "backend", 'proto/Python'))
+sys.path.insert(0, os.path.join(git_repo, "backend"))
```
Comment (Jgerbrandt) [inline backend/src/main.py:9]: Apparently git is not good for long running processes as it leaks resources. https://gitpython.readthedocs.io/en/stable/intro.html#limitations

What do we need this for anyway?

File: backend/src/SerialHandler.py (hunk 2 above)
Comment (EtSubas) [reply backend/src/SerialHandler.py]: I added this, needed to be veerified on hardware but software wise seems fine

File: backend/src/main.py (hunk 9 above)
Comment (EtSubas) [reply backend/src/main.py]: swapped for basic os operations

File: backend/src/SerialHandler.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    ser_han = SerialHandler(thread_name, port, baudrate, message_handler_workq)
+    if ser_han.serial_port == None:
+        return
+    
+    rx_thread = threading.Thread(target=serial_rx_thread, args=(ser_han,))
+    rx_thread.start()
+    while (1):
+        # If there is any workq messages, process them first
+        # then once the queue is empty read the serial port
+        if not serial_workq.empty():
```
Comment (cjchanx) [inline backend/src/SerialHandler.py:261]: Since we don't need to poll a separate thing now let's get rid of this empty and block forever for a workq item, this'll also give the rx thread more time to run.

**Review by cjchanx — APPROVED** [review]
Just gotta get rid of one more polling spot, otherwise looks good

**Review by Jgerbrandt — APPROVED** [review]
Address Chris' last comment and then i'm good.
